Add infrastructure for handling SPIR-V versions and extensions

AlexeySachkov opened the pull request:
Prototype of #110 

Sorry for such a huge PR, but I've tried to split into a smaller patches to simplify the review.

There are still some FIXMEs/TODOs in the code: I've spotted a few places which can be improved, a few issues with intel-specific extensions, etc. I don't want to fix all of these in this PR since it is huge enough already, but I will do so (or delegate to someone) as follow-up PRs.

For anyone who is not such interested in the implementation, please review at least the last patch `
Add documentation for version and extensions handling`

I would like also to write some kind of `Extending LLVM-SPIRV-Translator` document which will describe how to add support for new extensions to the translator: where to look in the code, which guards/checks for new instructions are required to be inserted, which testing coverage is expected, etc. Not sure that I will prepare this document anytime soon, but I will do it later as a follow-up PR.

Any feedback/comments/suggesstions is welcome and appreciated

AlexeySachkov commented:
@svenvh, @AlexeySotkin, @AnastasiaStulova, @neildhickey, @kpet, @pierremoreau, @airlied, @keryell, @Naghasan, @bader, @PawelJurek, @aratajew, @bashbaug,  @StuartDBrady, please review if you are interested.

@karolherbst, this PR will help to correctly implement #35, please review.

@oscarbg, this PR is related to some of your questions in #201, please review at least documentation for the new feature and send feedback

@bader, @PawelJurek, @aratajew, @vmaksimo, @MrSidims, please note that once this PR will be merged, additional arguments have to be passed to the translator in order to continue generating functionality from the intel-specific extensions, see corresponding commits in this PR.

karolherbst commented:
@AlexeySachkov The changes look fine in general, I am just concerned that it's overcomplicating things a little.

I don't think we need a "target" and a "max" version. Normally it should be enough to know the highest version the consumer supports and then we can either enable features as part of spir-v core or as a spir-v extension.

I am especially worried about the TranslatorOpts::isItAllowedToUseVersion method. So if we specify the target to be 1.3, why wouldn't we want to allow features valid for 1.1?

Right now I don't see any benefit having to deal with two fields adding constraints to what features the translator is allowed to emit.

This may also allow to get rid of VersionNumber::Undefined which seems to cause more issues than it actually helps. We could just default to 1.1 for eternity (or 1.2 or whatever makes sense) and depend on the consumer to tell the translator what it supports.

karolherbst commented:
speaking of extensions. It might makes sense to allow consumers to specify what spir-v extensions they support, so that the translator can enable features through extensions based on that.

AlexeySachkov commented:
@karolherbst, thanks for the review!

> I don't think we need a "target" and a "max" version. Normally it should be enough to know the highest version the consumer supports and then we can either enable features as part of spir-v core or as a spir-v extension.

That's actually a good point. If you allowed to use SPIR-V 1.2, then translator have to use features from it if input file requires to do so. In this context, target spir-v version looks unnecessary.

> I am especially worried about the TranslatorOpts::isItAllowedToUseVersion method. So if we specify the target to be 1.3, why wouldn't we want to allow features valid for 1.1?

If I understand correctly, SPIR-V spec is backward compatible and SPIR-V 1.3 includes 1.1 features. However, if some spec revision will break  backward compatibility, it might be useful to set target/minimum required SPIR-V version to avoid usage of older specs.

Anyway, let's wait for other reviews and opinions and then I will update the PR to reflect comments.

> It might makes sense to allow consumers to specify what spir-v extensions they support, so that the translator can enable features through extensions based on that.

Could you please explain this a little bit? Translator by itself knows nothing about where generated SPIR-V will be consumed and can only enable those extensions that were explicitly allowed by the user.

karolherbst commented:
> > It might makes sense to allow consumers to specify what spir-v extensions they support, so that the translator can enable features through extensions based on that.
> 
> Could you please explain this a little bit? Translator by itself knows nothing about where generated SPIR-V will be consumed and can only enable those extensions that were explicitly allowed by the user.

is there already an API for the user to tell the translator what extensions are supported and which not? I thought there is none and the translator right now just enables whatever it needs.

AlexeySachkov commented:
@karolherbst, 

> is there already an API for the user to tell the translator what extensions are supported and which not? I thought there is none and the translator right now just enables whatever it needs.

Yes, there are two ways of doing that:
1. If you are using `llvm-spirv` tool, then [--spirv-ext](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/c3fa9bf9fe20e12d7d48015c1913e0a06aeff24c) option is what you are looking for
2. If you are using translator as library, then [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/6c4b5090671f8b91cb33a2ba57bd531ac826e676#diff-c9e05e1657204d563a5aa4fa2a6882a3R120) and [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/5fb181f61d0d2047c5136baee473aba95e4d1514#diff-c9e05e1657204d563a5aa4fa2a6882a3R99) you can find new APIs which accept `TranslatorOpts` object. This object, in turn, [contains](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/c3fa9bf9fe20e12d7d48015c1913e0a06aeff24c#diff-54af6cfc9ddf4b456f94d7248c46b997R78) map for specifying which extension is enabled and which is not

karolherbst commented:
> 2. If you are using translator as library, then [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/6c4b5090671f8b91cb33a2ba57bd531ac826e676#diff-c9e05e1657204d563a5aa4fa2a6882a3R120) and [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/5fb181f61d0d2047c5136baee473aba95e4d1514#diff-c9e05e1657204d563a5aa4fa2a6882a3R99) you can find new APIs which accept `TranslatorOpts` object. This object, in turn, [contains](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/commits/c3fa9bf9fe20e12d7d48015c1913e0a06aeff24c#diff-54af6cfc9ddf4b456f94d7248c46b997R78) map for specifying which extension is enabled and which is not

cool, good to know that's also part of this PR :) would make the life easier for us inside mesa as we could then just whitelist extensions we know are working.

bashbaug commented:
Hi @AlexeySachkov , thank you for adding me to this review.

One of the usage models I'm interested in supporting longer-term is the ability to provide a clCreateProgramWithSource-like interface that invokes clang and the SPIR-V LLVM Translator, then passes the resulting SPIR-V module to the driver.  For example, this mechanism could be used to compile C++ for OpenCL in Clang from source, as-if it were supported natively in the driver.

Ideally, in this usage model, the layer providing the clCreateProgramWithSource interface wouldn't need to be updated for each new extension - it'd simply query which extensions are supported and pass them to Clang or the SPIR-V LLVM Translator.  With these proposed changes this is challenging though, since it requires passing the list of supported SPIR-V extensions to the SPIR-V LLVM Translator and there is no way to query a "SPIR-V extension string" at the moment.

Is there a way to reduce the amount of information that needs to be passed to the SPIR-V LLVM Translator, or to restrict the information we need to pass to queryable information?  (Should we request to add a query for supported SPIR-V extensions?)

AlexeySachkov commented:
Hi @bashbaug

> Ideally, in this usage model, the layer providing the clCreateProgramWithSource interface wouldn't need to be updated for each new extension - it'd simply query which extensions are supported and pass them to Clang or the SPIR-V LLVM Translator. 

It should be easy to add a query to the SPIR-V LLVM Translator to get all **known** extensions. Moreover, some kind of such query is already available in this PR: any tool that wants to get list of known extension can include [LLVMSPIRVExtensions.inc](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/09ff9bb4c758cac8c4f7ff4fc4c5d3696d0110dd/include/LLVMSPIRVExtensions.inc) and use it to populate any data structure it wants. See examples [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/files#diff-54af6cfc9ddf4b456f94d7248c46b997R60) and [here](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244/files#diff-8b7efcbcbffdd234c43aeda3bb73eaa5R291)

But if you want to get list of extensions which needs to be enabled for this particular LLVM IR input file...not sure that it can be fully automated.

First of all, you could just enable **all** known extensions in the translator, but I'm aware that it is not the best way. For example, there are [SPV_KHR_shader_ballot](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_shader_ballot.html) and [SPV_AMD_shader_ballot](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/AMD/SPV_AMD_shader_ballot.html) extensions - which one should be used if both of them enabled? 

I think some kind of mapping between clang and SPIR-V extensions could help you there: you will do then something like:
```c++
  performClangStep();
  ExtensionsToEnable = mapToSPIRVExtensions(getUsedOpenCLExtensions());
  // adjust ExtensionsToEnable if needed unconditionally or conditionally based on other info
  translateSPIRVToLLVMIR(ExtensionsToEnable);
```

AlexeySachkov commented:
Hi @tremmelg, thanks for the review!

> I would think just having the target version is enough. Some features are missing from later versions of the SPIR-V specification.

Not sure that it is good idea, especially during consumption step, see [this comment](https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/244#issuecomment-510007705) from @karolherbst:

> I am especially worried about the TranslatorOpts::isItAllowedToUseVersion method. So if we specify the target to be 1.3, why wouldn't we want to allow features valid for 1.1?

Right now we have `max` version specifically for this purposes.

I was thinking about the following:

* Let's have two fields in `TranslatorOpts`: one for minimum allowed version and another one for maximum allowed version. When you are using translator as library, you can configure it to allow usage of only one particular version (via `max` = `min` = whatever you want) or of a range of version. By default this versions are equal to the lowest and the highest known SPIR-V versions. This will help to get rid of `Undefined` SPIR-V version.
* Let's have **three** command line arguments for other users: two for `min` and `max` versions and the third one for `target` version which will be alias to the combination of first ones.

I'm not particularly sure how important is third option (`target` version), but `min` and `max` looks required to be able to handle most of the cases.

@karolherbst, @tremmelg, @AnastasiaStulova, @AlexeySotkin, what do you think?

karolherbst commented:
What are the benefits of having a min version specified? Just to be future proof or are there any technical reasons as well?

Hopefully, if spir-v continues to evolve like it already does, a consumer should always be a able to consume a spir-v only using features from older versions.

Except features get removed, .. so maybe that would be a valid reason.

But I think the target isn't something we really need. Min would protect the consumer against already removed features from spir-v and the Max one against newer features the consumer doesn't support yet.

But.. I wouldn't know what the target version should affect. Like what should the translator do if min=1.1 max=1.4 target=1.3?

tremmelg commented:
As I pointed out, some features have been removed.  For example, OpAtomicCompareExchangeWeak is missing after SPIR-V 1.3 (https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_atomic_a_atomic_instructions).

AlexeySachkov commented:
> But.. I wouldn't know what the target version should affect. Like what should the translator do if min=1.1 max=1.4 target=1.3?

I think if `target` option was specified, `max` and `min` cannot be specified anymore - these two are different kinds of controls.

From my point of view `target` looks like some simple alias if you want to experiment with certain SPIR-V version. Otherwise, `min`/`max` is easier for you, for example:
* you only support SPIR-V 1.2 and below: `-spirv-max-version=1.2`
* you want to stick with the latest specs: `-spirv-min-version=1.4`

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+### Handling SPIR-V versions generated by the translator
+
+There are two options to control behavior of the translator around version of
+SPIR-V file which is generated/consumed.
```
Fixed, thanks!

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+There are two options to control behavior of the translator around version of
+SPIR-V file which is generated/consumed.
+
+* `-spirv-target-version=` - this option allows to force SPIRV-LLVM-Translator
```
I've removed `target` option completely

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+  to generate SPIR-V file corresponding to particular specified version of the
+  specification.
+
+  If `-r` option was also specified, SPIRV-LLVM-Translator will reject input
```
I've removed `target` option completely

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+  file and emit error if SPIR-V version in it doesn't equal to the specified one
+  via this option.
+
+* `-spirv-max-version=` - this option allows to restrict SPIRV-LLVM-Translator
```
Fixed, thanks!

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+  to generate SPIR-V file which correspond to a version of the specification
+  which is higher than specified one via this option.
+
+  If `-r` option was also specified, SPIRV-LLVM-Translator will reject input
```
Fixed, thanks!

AlexeySachkov commented on include/LLVMSPIRVOpts.h:
```diff
@@ -0,0 +1,152 @@
+enum class VersionNumber : uint32_t {
+  // See section 2.3 of SPIR-V spec: Physical Layout of a SPIR_V Module and
+  // Instruction
+  Undefined = 0x0,
```
Yep, done

AlexeySachkov commented on include/LLVMSPIRVOpts.h:
```diff
@@ -0,0 +1,152 @@
+                 const ExtensionsStatusMap &Map = {}) {
+    assert((VersionNumber::Undefined == Target ||
+            VersionNumber::Undefined == Max) &&
+           "Inconsitent state");
```
Not actual anymore

AlexeySachkov commented on include/LLVMSPIRVOpts.h:
```diff
@@ -0,0 +1,152 @@
+  getWithTargetVersion(VersionNumber Target,
+                       const ExtensionsStatusMap &Map = {}) {
+    TranslatorOpts Result;
+    Result.setTargetVersion(Target);
```
Not actual anymore

AlexeySachkov commented on include/LLVMSPIRVOpts.h:
```diff
@@ -0,0 +1,152 @@
+    return Result;
+  }
+
+  bool isItAllowedToUseVersion(VersionNumber AskedVersion) const {
```
Done

AlexeySachkov commented on include/LLVMSPIRVLib.h:
```diff
@@ -93,6 +96,12 @@ bool isSpirvText(std::string &Img);
+/// \brief Load SPIR-V from istream as a SPIRVModule.
+/// \returns null on failure.
+std::unique_ptr<SPIRVModule> readSpirvModule(std::istream &IS,
+                                             SPIRV::TranslatorOpts &Opts,
```
Done

AlexeySachkov commented on lib/SPIRV/libSPIRV/SPIRVModule.cpp:
```diff
@@ -1552,9 +1565,19 @@ std::istream &operator>>(std::istream &I, SPIRVModule &M) {
+  if (!M.getErrorLog().checkError(
+          MI.SPIRVVersion <=
+              static_cast<uint32_t>(VersionNumber::MaximumVersion),
+          SPIRVEC_InvalidModule, "unsupported SPIR-V version number")) {
```
Done

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+
+### Handling SPIR-V extensions generated by the translator
+
+By default, translator doesn't use any extensions. If it required to enable
```
Fixed, thanks!

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+By default, translator doesn't use any extensions. If it required to enable
+certain extension, the following command line option can be used:
+
+* ``--spirv-ext=`` - allows to control list of allowed/disallowed extensions.
```
Fixed, thanks!

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+name in this option: ``all`` - it affects all extension which are known to the
+translator.
+
+If ``--spirv-ext`` contains name of extension which is not know for the
```
Fixed, thanks!

AlexeySachkov commented on README.md:
```diff
@@ -97,6 +97,54 @@ To translate between LLVM IR and SPIR-V:
+If ``--spirv-ext`` contains name of extension which is not know for the
+translator, it will emit error.
+
+If `-r` option was specified, translator will accept SPIR-V that uses any known
```
Updated

AlexeySachkov commented on include/LLVMSPIRVOpts.h:
```diff
@@ -0,0 +1,152 @@
+      : TargetVersion(VersionNumber::Undefined),
+        MaxVersion(VersionNumber::MaximumVersion) {}
+
+  TranslatorOpts(VersionNumber Target, VersionNumber Max,
```
Not actual anymore

AlexeySachkov commented on lib/SPIRV/libSPIRV/SPIRVEntry.cpp:
```diff
@@ -424,6 +424,15 @@ void SPIRVEntry::updateModuleVersion() const {
   Module->setMinSPIRVVersion(getRequiredSPIRVVersion());
 }
 
+bool SPIRVEntry::isItAllowedToUseExtension(ExtensionID AskedExtension) const {
```
This is more like a helper, but since you can always access to `Module` from any sub-class of `SPIRVEntry` I decided to remove them to avoid possible confusion

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+   means that ``+all,`` is implicitly added at the beggining of the
+   ``-spirv-ext`` value.
+
+This behavior, however, can be controlled via the same switches described in
```
Updated

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+
+By default, translator consumes SPIR-V of any version which is supported.
+
+This behavior, however, can be controlled via the same switches described in
```
Fixed

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+If any particular SPIR-V version is required to be generated, behavior of the
+translator can be controlled by the following two switches:
+
+* ``--spirv-target-version=`` - instructs the translator to generate SPIR-V file
```
Fixed, thanks!

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+   submit pull request - contributions are welcome!
+
+If any particular SPIR-V version is required to be generated, behavior of the
+translator can be controlled by the following two switches:
```
Fixed, thanks!

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+
+.. note::
+   There is no documentation about which exact features from newest
+   SPIR-V spec versions will be used by the translator - this fact should be
```
Updated

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+the consumer to analyze incoming SPIR-V file and reject it if it contains
+something that is not supported by the consumer.
+
+However, translator could simplify this step for downstream users by checking
```
Updated

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+``SPV_KHR_no_integer_wrap_decoration`` - it will be always generated if
+corresponding LLVM IR counterparts are encountered in input module.
+
+It worth mentioning that SPIR-V versions and extensions handling is mostly
```
Fixed, thanks!

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+SPIR-V according to the core spec, without uses of any extensions if such SPIR-V
+was requested by user.
+
+For example, without such infrastructure it is impossible to disable use of
```
Fixed, thanks!

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+must be possibility to control which SPIR-V spec will be used during
+translation.
+
+SPIR-V extensions is another one thing which must be controllable. Extensions
```
Fixed, thanks!

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+This document describes how translator makes decision about using instructions
+from different version of SPIR-V core and extension specifications.
+
+Being able to control resulting SPIR-V version is an important thing: target
```
Fixed, thanks!

AlexeySachkov commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,255 @@
+Overview
+========
+
+This document describes how translator makes decision about using instructions
```
Fixed, thanks!

AlexeySachkov commented:
Hi there,

I've updated the PR:
* rebased to the latest `master`
* added a few more commits fixing LIT tests related to newly added features from different SPIR-V extensions
* applied comments from @tremmelg 
* removed `--spirv-target-version` command line option since it doesn't seem desired

I haven't added `--spirv-min-version` command line option and corresponding functionality, since it requires some additional work - I think since the translator doesn't support SPIR-V 1.4 yet it is not so critical.

At least I would like to do so in a separate PR since this one is huge enough already. Also, it would be nice to get it merged soon because people add support for new SPIR-V extensions constantly and it takes additional time to update their code to new infrastructure

AlexeySachkov commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -2005,10 +2005,6 @@ bool llvm::writeSpirv(Module *M, const SPIRV::TranslatorOpts &Opts,
 }
 
 bool llvm::regularizeLlvmForSpirv(Module *M, std::string &ErrMsg) {
-  std::unique_ptr<SPIRVModule> BM(SPIRVModule::createSPIRVModule());
```
@svenvh, could you please pay attention to this particular change?

When I started working on this PR these lines looked like:
```c++
  std::unique_ptr<SPIRVModule> BM(SPIRVModule::createSPIRVModule());
```

And `BM` was unused, so - I deleted it. Right now it looks like `BM` is used to emit error messages and probably it is better to leave this code as-is

svenvh replied:
Yes it is better to keep `BM` as we use the error log to report any issues found by `isValidLLVMModule`; you are (unintentionally?) deleting that check actually.

AlexeySachkov replied:
Thanks, I will leave this code as-is

AlexeySachkov replied:
Restored this code and added test for it

AlexeySachkov commented on test/spirv-version-controls.ll:
```diff
@@ -0,0 +1,33 @@
+; RUN: llvm-as %s -o %t.bc
+;
+; For now this is dummy tests to check that new options are consumed
+; RUN: llvm-spirv %t.bc -spirv-text --spirv-max-version=1.0 -o %t
```
I have doubts that this test is really useful. I can expand it to test that resulting SPIR-V version will not be greater than specified by `--spirv-max-version`, but then it will look very similar to test for `DecorationMaxByteOffset`. Any other opinions?

svenvh replied:
Probably not that useful then.  I'd expect that new tests for new SPIR-V features will make more use of the `--spirv-max-version` option anyway.

AlexeySachkov replied:
Thanks, I will remove it then

AlexeySachkov commented:
Fixed clang-format issues. Updated existing `readSpirv` and `wirteSpirv` APIs to preserve previous behavior around SPIR-V extensions (all enabled by default). Updated documentation to reflect the latest changes

AlexeySotkin commented on tools/llvm-spirv/llvm-spirv.cpp:
```diff
@@ -93,6 +96,22 @@ static cl::opt<bool>
+               clEnumValN(VersionNumber::SPIRV_1_1, "1.1", "SPIR-V 1.1")),
+    cl::init(VersionNumber::MaximumVersion));
+
+static cl::opt<std::string>
```
Should we use `cl::list` instead, like [here](https://github.com/llvm/llvm-project/blob/4f1d893f9ec33a493d9f35426254a0ce69716933/llvm/include/llvm/CodeGen/CommandFlags.inc#L38-L41)

AlexeySachkov replied:
Done, thanks for suggestion!

AlexeySotkin commented on test/spirv-extensions-control.spt:
```diff
@@ -0,0 +1,43 @@
+1 FunctionEnd
+
+; RUN: llvm-spirv %s -to-binary -o %t.spv
+; RUN: llvm-spirv -r %t.spv -o %t.bc
```
What is the purpose of this test ?

AlexeySachkov replied:
I've left a comment in the test

AlexeySotkin commented on lib/SPIRV/libSPIRV/SPIRVModule.cpp:
```diff
@@ -1589,9 +1610,26 @@ std::istream &operator>>(std::istream &I, SPIRVModule &M) {
+      MI.SPIRVVersion <= static_cast<uint32_t>(VersionNumber::MaximumVersion);
+  if (!M.getErrorLog().checkError(
+          SPIRVVersionIsKnown, SPIRVEC_InvalidModule,
+          "unknown SPIR-V version number. Range of supported/known SPIR-V "
```
I think it might be useful to know what the actual version is. Please print value of `MI.SPIRVVersion` to the error message.

AlexeySachkov replied:
Done

AlexeySotkin commented on tools/llvm-spirv/llvm-spirv.cpp:
```diff
@@ -259,13 +278,110 @@ static int regularizeLLVM() {
+      ExtensionsStatus[It->second] = ExtStatus;
+    }
+
+    // Check for +ext,-ext pairs
```
Should we have implementation aligned with -cl-ext ? If I recall correctly, for -cl-ext, if the same extension is encountered multiple times with possibly different signs the latter one takes precedence.

AlexeySachkov replied:
Done

AlexeySachkov commented:
Applied comments from @AlexeySotkin

svenvh commented on include/LLVMSPIRVOpts.h:
```diff
@@ -0,0 +1,104 @@
+//
+//                     The LLVM Compiler Infrastructure
+//
+// This file is distributed under the University of Illinois Open Source
```
As this is a new file, would it make sense to update the license header / copyright line to the new ones? (see #80 and #134).

AlexeySachkov replied:
I'm not sure about this particular one: I'm not very familiar with all this licensing stuff, but this file is going to be shipped together with the translator for integration with other libraries. That effectively means that downstream users that use pre-built packages will get a package with two different licenses - is it okay?

svenvh replied:
No idea to be honest; perhaps easier to leave it as is for now then.

StuartDBrady commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,247 @@
+* emits error if LLVM IR entity cannot be ignored and cannot be emulated using
+  available instructions.
+
+  **TODO**: Example
```
Is this example difficult to provide? Perhaps we should discuss this before merging?

AlexeySachkov replied:
If I remember correctly, @AnastasiaStulova proposed that if translator encounters some built-in call and it knows that this built-in comes from an extension, then it is better to emit error if corresponding SPIR-V extension wasn't enabled (I originally proposed to translate it as usual call of user's function).

I guess such situation can be added here as an example, what do you think?

AlexeySachkov replied:
Oh, this is about versions, not extensions.

Then, probably some static constructors/destructors might be an example here, but I'm not sure, need to check

AlexeySachkov replied:
I've added an example with global constructors/destructors

StuartDBrady commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,247 @@
+  easily in presence of some extension, but can be legally emulated without it.
+  I.e. in situations similar to ``OpPtrEqual`` in SPIR-V less than 1.4
+
+  **TODO:** add example if such exists.
```
Do you need help verifying this?

AlexeySachkov replied:
@StuartDBrady, if you have some examples or know how to easily get such, please let me know, it would be great

AlexeySachkov replied:
It turned out that it is not so easy to find some extension functionality which can be easily emulated with core functionality (what is the point of the extension then?)

The only one thing that I've found so far is  that `OpSubgroupAllKHR` instruction from [SPV_KHR_subgroup_vote](http://htmlpreview.github.io/?https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/KHR/SPV_KHR_subgroup_vote.html) can be replaced with `OpGroupAll` from the core SPIR-V spec. But, I haven't verified yet that `OpGroupAll` is available in the older SPIR-V versions, i.e. it is legal to replace one for another or not

AlexeySachkov replied:
So, I haven't found any good examples, so, I will remove this TODO. If we find something good to place here we can always update documentation later

svenvh commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,247 @@
+* emits error if LLVM IR entity cannot be ignored and cannot be emulated using
+  available instructions.
+
+  **TODO**: Example
```
Are you planning to fill in the TODO before merging this PR?
(and there are actually other TODOs in this PR too)

AlexeySachkov replied:
Here is the list of TODOs/FIXMEs in my PR with my plans around them:

* examples in documentation - will fix them in this PR
* LLVMSPIRVOpts.h:53 - not going to fix in this PR. Should be fixed when we update translator to support SPIR-V 1.2 and higher
* SPIRVValue - not going to fix in this PR. Should be fixed when we update the translator to support SPIR-V 1.4
* the rest are related to IntelFPGA-specific extensions and I also don't want to increase this PR with tests and translator improvements - let's say this PR introduces basic infrastructure and fixes LIT tests, but the rest of improvements that can be built on top of this infrastructure will go as separate PRs

StuartDBrady commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -1161,6 +1160,22 @@ parseAnnotations(StringRef AnnotatedCode) {
 void addIntelFPGADecorations(
     SPIRVEntry *E,
     std::vector<std::pair<Decoration, std::string>> &Decorations) {
+  // FIXME: move this code out of this function, it is not IntelFPGA-specific
```
Would this be easily resolved now?

AlexeySachkov replied:
I've refactored code related to IntelFPGA-specific decorations and removed this FIXME

StuartDBrady commented on lib/SPIRV/SPIRVWriter.cpp:
```diff
@@ -1199,16 +1214,29 @@ void addIntelFPGADecorations(
 void addIntelFPGADecorationsForStructMember(
     SPIRVEntry *E, SPIRVWord MemberNumber,
     std::vector<std::pair<Decoration, std::string>> &Decorations) {
+  // FIXME: move this code out of this function, it is not IntelFPGA-specific
```
Ditto?

AlexeySachkov replied:
I've refactored code related to IntelFPGA-specific decorations and removed this FIXME

StuartDBrady commented on tools/llvm-spirv/llvm-spirv.cpp:
```diff
@@ -259,13 +279,82 @@ static int regularizeLLVM() {
+  // Set the initial state
+  // During SPIR-V consumption, assume that any known extension is allowed
+  // During SPIR-V generation, assume that any known extension is disallowed
+  // During convertation to/from SPIR-V text representation, assume that any
```
convertation -> conversion?

svenvh replied:
And while you're here, probably worth ending sentences with `.` .

AlexeySachkov replied:
Fixed

StuartDBrady commented on tools/llvm-spirv/llvm-spirv.cpp:
```diff
@@ -259,13 +279,82 @@ static int regularizeLLVM() {
   cl::ParseCommandLineOptions(Ac, Av, "LLVM/SPIR-V translator");
 
+  SPIRV::TranslatorOpts::ExtensionsStatusMap ExtensionsStatus;
+  // ExtensionsStatus will be properly initialized and update accrording to
```
accrording -> according

AlexeySachkov replied:
Fixed, thanks!

StuartDBrady commented on test/spirv-unknown-extensions.spt:
```diff
@@ -0,0 +1,44 @@
+119734787 65536 393230 16 0 
```
There's stray whitespace at the ends of many of the lines in this test. If you push a new version, it would be good to fix this.

AlexeySotkin replied:
I believe those whitespaces were inserted by `llvm-spirv` tool.

AlexeySachkov replied:
Fixed

StuartDBrady commented on test/spirv-version-controls-negative-1.spt:
```diff
@@ -0,0 +1,33 @@
+119734787 66560 393230 12 0·
```
There's unusual punctuation at the ends of each of the lines in this test. Presumably this was from a file that rendered spaces at the ends of lines as "·". It would be good to remove this.

AlexeySachkov replied:
Yes, you are right about rendering trailing whitespaces as "·" - will remove this

AlexeySachkov replied:
Fixed

StuartDBrady commented on test/spirv-version-controls-negative-2.spt:
```diff
@@ -0,0 +1,34 @@
+119734787 1024 393230 12 0·
```
Again, there is unusual punctuation at the ends of the lines in this test, which would be good to remove.

AlexeySachkov replied:
Fixed, thanks!

AlexeySachkov commented:
Applied comments from @svenvh and @StuartDBrady

@svenvh, 

> As this is a large PR: does every commit pass the lit tests? If not we should probably do some squashing to ensure that we keep the master branch bisectable.

Here is the list of patches with number of LIT tests failed per each commit:
```
[NFC] Extend translator API with translation options          | 0 
Add command line options to restrict max SPIR-V version       | 0 
Propagate TranslationOpts to SPIRVModule                      | 0 
Remove -spirv-no-deref-attr option in favor of new options    | 0 
Add handling of max SPIR-V version to SPIRVReader             | 0 
Extend TranslatorOpts to handle SPIR-V extensions             | 0 
Update extensions handling infrastucture                      | 46
Handle extensions during SPIR-V consumption                   | 46
Correctly handle SPV_KHR_no_integer_wrap_decoration           | 10
Fix several LIT tests related to llvm.annotate intrinsics     | 8 
Fix SPV_INTEL_device_side_avc_motion_estimation LIT tests     | 4 
Fix LIT tests related to SPV_INTEL_fpga_reg extension         | 3 
Fix LIT tests related to SPV_INTEL_blocking_pipes extension   | 2 
Fix LIT tests related to SPV_INTEL_unstructured_loop_controls | 0 
Add documentation for version and extensions handling         | 0 
```

Note here: I'm testing with assertions enabled and all failed LIT tests are failed due to assertions. I think it worth squashing several patches with fixes - I can do that before merging once we are done with review.

For now, if there are no other objections, the only leftover is to add one more example into documentation (marked as `TODO`)

StuartDBrady commented on test/spirv-version-controls-negative-1.spt:
```diff
@@ -0,0 +1,33 @@
+3 MemoryModel 1 2
+4 EntryPoint 6 6 "foo"
+9 String 11 "kernel_arg_type.foo.int*,"
+3 Source 3 102000·
```
There's still one left on this line.

AlexeySachkov replied:
Fixed

AlexeySachkov commented:
Fixed clang-format issue, comment from @StuartDBrady  and rebased to the latest master to get fix for "enqueue" tests.

@svenvh, I've re-structured patches a little bit, so the status of LIT tests looks like:
```
Update tests for LLVM unnamed argument printing               |
[NFC] Extend translator API with translation options          |
Add command line options to restrict max SPIR-V version       |
Propagate TranslationOpts to SPIRVModule                      |
Remove -spirv-no-deref-attr option in favor of new options    |
Add handling of max SPIR-V version to SPIRVReader             |
Extend TranslatorOpts to handle SPIR-V extensions             |
Update extensions handling infrastucture                      |
Add an assertion to check correctness of created module       |   Unexpected Failures: 46
Correctly handle SPV_KHR_no_integer_wrap_decoration           |   Unexpected Failures: 10
Fix several LIT tests related to llvm.annotate intrinsics     |   Unexpected Failures: 8
Fix SPV_INTEL_device_side_avc_motion_estimation LIT tests     |   Unexpected Failures: 4
Fix LIT tests related to SPV_INTEL_fpga_reg extension         |   Unexpected Failures: 3
Fix LIT tests related to SPV_INTEL_blocking_pipes extension   |   Unexpected Failures: 2
Fix LIT tests related to SPV_INTEL_unstructured_loop_controls |
Add documentation for version and extensions handling         |
```

There is bunch of "clean" patches, then one patch that brings assertions and causes a lot of regressions - then bunch of patches with fixed. So, I can squash one small patch about assertions and all patches with fixes into one before merging.

TODO in the documentation is still in place, I've commented there that my first attempt to find an example failed - I will make another one and will remove it as "no examples found" if the second attempt also fails

AlexeySotkin commented on docs/SPIRVVersionsAndExtensionsHandling.rst:
```diff
@@ -0,0 +1,250 @@
+   broke workflows of existing users of the translator.
+
+.. note::
+   Since by default during SPIR-V consumption all extensions are disabled, this
```
> all extensions are **disabled**

Looks like it should be *enabled*, right ?

AlexeySachkov replied:
Yes, indeed. Fixed

AlexeySachkov commented:
Applied comment from @AlexeySotkin and removed one TODO from documentation since I wasn't able to find any examples.

@svenvh, @StuartDBrady, could you please let me know if you have any other comments/feedback about this PR?
If it is good enough to be accepted, then I will squash several patches together to ensure that each commit passes LIT tests and we can merge it then.

AlexeySachkov commented:
> I am concerned though that squash these commits too much will make it hard to port to the llvm_release_80 and llvm_release_70 branches which lack support for vendor-specific extensions.

@StuartDBrady, fair point.

My current plan is to squash only these commits:
```
Add an assertion to check correctness of created module       |   Unexpected Failures: 46
Correctly handle SPV_KHR_no_integer_wrap_decoration           |   Unexpected Failures: 10
Fix several LIT tests related to llvm.annotate intrinsics     |   Unexpected Failures: 8
Fix SPV_INTEL_device_side_avc_motion_estimation LIT tests     |   Unexpected Failures: 4
Fix LIT tests related to SPV_INTEL_fpga_reg extension         |   Unexpected Failures: 3
Fix LIT tests related to SPV_INTEL_blocking_pipes extension   |   Unexpected Failures: 2
Fix LIT tests related to SPV_INTEL_unstructured_loop_controls |
```

However, if our first priority is buildable state and having all LIT tests passed on each commit is less important, then we can avoid squashing at all.@svenvh, what do you think?

svenvh commented:
> > I am concerned though that squash these commits too much will make it hard to port to the llvm_release_80 and llvm_release_70 branches which lack support for vendor-specific extensions.

Any idea how hard that would really be?

> However, if our first priority is buildable state and having all LIT tests passed on each commit is less important, then we can avoid squashing at all.@svenvh, what do you think?

I still have a strong preference for the history to be bisectable using LIT tests.

Alternatively, can't you just reorder your commits such that the commit breaking the tests comes after the commit(s) fixing the tests?

AlexeySachkov commented:
@svenvh,

> Any idea how hard that would really be?

I think the main problem would be that patch won't apply at all if some of recently added extensions aren't supported at a release branch

> Alternatively, can't you just reorder your commits such that the commit breaking the tests comes after the commit(s) fixing the tests?

Great idea, let me try it

AlexeySachkov commented:
@svenvh, I've re-ordered patches, so, now each commit is buildable and all LIT tests pass.

I also updated commit messages a little bit to reflect new order of patches

karolherbst commented:
mind adding some spirv-val checks to some of the tests? Especially the spt ones so that we make sure we only test with valid spirv files, otherwise somebody tries to track down a test regression caused by invalid spirv.

(We really should add it to all tests, but there were some annoying issues and I don't really have that much time to look into those :/ )

AlexeySachkov commented:
@karolherbst,

> mind adding some spirv-val checks to some of the tests? 

I've updated two `.spt` tests to launch `spirv-val`. The rest of affected tests are related to some extension that cannot be validated by this tool and some of new `.spt` files were made invalid on purpose, because they are negative tests.

AlexeySachkov commented:
Thanks everyone who participated for the review!

If you spot any issues - let me know by tagging me (@AlexeySachkov)
